openglyph/src/io/mega_file.cpp

```diff
@@ -0,0 +1,162 @@
+#include <openglyph/io/mega_file.hpp>
+
+namespace openglyph::io {
+
+class MegaFile::SubFile : public khepri::io::Stream
+{
+public:
+    SubFile(const SubFileInfo info, khepri::io::File* p_file);
+
+    SubFile(const SubFile&)            = delete;
+    SubFile(SubFile&&)                 = delete;
+    SubFile& operator=(const SubFile&) = delete;
+    SubFile& operator=(SubFile&&)      = delete;
+
+    /// \see stream::readable
+    [[nodiscard]] bool readable() const noexcept override
+    {
+        return true;
+    }
+
+    /// \see stream::writable
+    [[nodiscard]] bool writable() const noexcept override
+    {
+        return false;
+    }
+
+    /// \see stream::seekable
+    [[nodiscard]] bool seekable() const noexcept override
+    {
+        return true;
+    }
+
+    /**
+     * \brief Reads data from the stream at the current position.
+     *
+     * \param[in] buffer the memory to write the read data to.
+     * \param[in] count the number of consecutive bytes to read.
+     *
+     * \return the number of bytes read and stored in \a buffer.
+     */
+    virtual size_t read(void* buffer, size_t count) override;
+
+    /**
+     * \brief Writes data to the stream at the current position (not supported for subfiles).
+     *
+     * \param[in] buffer pointer to the data to write.
+     * \param[in] count the number of consecutive bytes to write.
+     *
+     * \return the number of bytes from \a buffer written to the stream.
+     */
+    virtual size_t write(const void* buffer, size_t count) override;
+
+    /**
+     * \brief Changes the file position
+     *
+     * \param[in] offset offset by how much to change the file position
+     * \param[in] origin the origin from whence to change the position
+     *
+     * \return the new file position, from the start of the file
+     */
+    virtual long long seek(long long offset, khepri::io::SeekOrigin origin) override;
+
+private:
+    const SubFileInfo info;
+    khepri::io::File* p_megaFile;
+    std::uint64_t     local_read_offset;
+};
+
+MegaFile::MegaFile(const std::filesystem::path& mega_file_path)
+    : m_megaFile(std::make_unique<khepri::io::File>(mega_file_path, khepri::io::OpenMode::read))
+{
+    extract_metadata(filenames, fileinfo);
+
+    // calculate hashes for this session
+    for (const SubFileInfo& info : fileinfo) {
+        std::size_t hash = std::hash<std::string>{}(filenames[info.file_name_index]);
+        file_hashes.push_back(hash);
+    }
+}
+
+std::unique_ptr<khepri::io::Stream> MegaFile::open_file(std::filesystem::path& path)
+{
+    std::size_t hash = std::hash<std::string>{}(khepri::uppercase(path.string()));
+    auto        it   = std::find(file_hashes.begin(), file_hashes.end(), hash);
+
+    if (it != file_hashes.end()) {
+        std::size_t index = static_cast<std::size_t>(std::distance(file_hashes.begin(), it));
+        return std::make_unique<SubFile>(fileinfo[index], m_megaFile.get());
+    }
+    throw khepri::io::FileNotFoundError();
```

Comment:
This isn't an error, just return `nullptr`. It can happen quite frequently with modpaths.